<div align="center">
<h1>Deep learning Practical Assignment 2:
<h6>Linear Classifier : Fully connected networks
<h4>Lab date : October 25th, 2021
<h4>Deadline to submit : November 25th, 2021 - 11:59 PM (23:59)   
</div>
<h3>Group Members: <ul>
    <li> Ahmed Belaaj
    <li> Omar Chaabouni
    <li> Linna Azaiez
    </ul> 

## Installing required packages :

In [1]:
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
import warnings
warnings.filterwarnings('ignore')

## 1 Implementing fully connected dataset !!
### 1.1 Toy data set

In [2]:
def prepare_data( var_1 , var_2 ):
    mean1= [-1 , 0]
    mean2= [1 , 0]
    cov_1 = var_1 * np.eye(2) 
    cov_2 = var_2 * np.eye(2)
    x = np.concatenate( (np.random.multivariate_normal(mean1, cov_1, 125) , np.random.multivariate_normal(mean2, cov_2, 125))  )
    y = np.concatenate((np.ones(125),-1*np.ones(125)))
    x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.20, shuffle = True )
    return x_train, x_test, y_train, y_test 

In [3]:
# we set var_1 = var_2 = 0.75
var_1 = 0.75
var_2 = 0.75
x_train, x_test, y_train, y_test = prepare_data( var_1 , var_2 ) 

### 1.2 Using keras

In [4]:
from keras.models import Sequential
from keras.layers import Dense ,Dropout
from keras.utils import to_categorical

model = Sequential()
model.add(Dense(8, activation="relu", input_shape=(2,)))
model.add( Dense (5 , activation ='relu'))
model.add( Dense (2 , activation ='tanh'))

model.compile ( loss ="binary_crossentropy", optimizer ='adam' , metrics =['accuracy'])
model.fit (x_train , to_categorical(y_train) ,  epochs =15 , batch_size =10 ,verbose =2)

Epoch 1/15
20/20 - 0s - loss: 7.9402 - accuracy: 0.0900
Epoch 2/15
20/20 - 0s - loss: 7.8788 - accuracy: 0.2450
Epoch 3/15
20/20 - 0s - loss: 7.2283 - accuracy: 0.3150
Epoch 4/15
20/20 - 0s - loss: 5.6037 - accuracy: 0.4150
Epoch 5/15
20/20 - 0s - loss: 4.7056 - accuracy: 0.5350
Epoch 6/15
20/20 - 0s - loss: 4.2203 - accuracy: 0.5950
Epoch 7/15
20/20 - 0s - loss: 3.9081 - accuracy: 0.6800
Epoch 8/15
20/20 - 0s - loss: 3.5573 - accuracy: 0.7250
Epoch 9/15
20/20 - 0s - loss: 2.8498 - accuracy: 0.7950
Epoch 10/15
20/20 - 0s - loss: 2.0451 - accuracy: 0.8950
Epoch 11/15
20/20 - 0s - loss: 1.8567 - accuracy: 0.9350
Epoch 12/15
20/20 - 0s - loss: 1.5832 - accuracy: 0.9650
Epoch 13/15
20/20 - 0s - loss: 1.2984 - accuracy: 0.9800
Epoch 14/15
20/20 - 0s - loss: 1.0954 - accuracy: 0.9900
Epoch 15/15
20/20 - 0s - loss: 0.9480 - accuracy: 0.9950


## 2. The IRIS Dataset

In [5]:
from sklearn.datasets import load_iris
from keras.metrics import Precision , Recall
(X_iris , y_iris) = load_iris(return_X_y=True , as_frame=True)
y_iris = to_categorical(y_iris)
x_train, x_test, y_train, y_test = train_test_split(X_iris , y_iris, test_size=0.20, shuffle = True , stratify = y_iris )

In [6]:
input_dim = x_train.shape[1]
output_dim = y_test.shape[1]

In [7]:
def create_custom_model(input_dim, output_dim, nodes, n , loss , name='model'):
    def create_model():
        # Create model
        model = Sequential(name=name)
        model.add(Dense(nodes, input_shape=(input_dim,) , activation='relu'))
        for i in range(n-1):
            model.add(Dense(nodes, activation='relu'))
        model.add(Dense(output_dim, activation='softmax'))

        # Compile model
        model.compile(loss=loss, 
                      optimizer='adam', 
                      metrics=['accuracy',Precision(),Recall()])
        return model
    return create_model

models = [create_custom_model(input_dim, output_dim, 256, 1, "categorical_crossentropy" , 'model_1'),
          create_custom_model(input_dim, output_dim, 256, 1, "kl_divergence" , 'model_2'),
          create_custom_model(input_dim, output_dim, 128, 2, "categorical_crossentropy" ,'model_3' ),
          create_custom_model(input_dim, output_dim, 128, 2, "kl_divergence" , 'model_4')
          ]

In [8]:
for model in models:
    model().summary()

Model: "model_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_3 (Dense)              (None, 256)               1280      
_________________________________________________________________
dense_4 (Dense)              (None, 3)                 771       
Total params: 2,051
Trainable params: 2,051
Non-trainable params: 0
_________________________________________________________________
Model: "model_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_5 (Dense)              (None, 256)               1280      
_________________________________________________________________
dense_6 (Dense)              (None, 3)                 771       
Total params: 2,051
Trainable params: 2,051
Non-trainable params: 0
_________________________________________________________________
Model: "model_3"
_____________________

## IRIS dataset
### Experiment 1 :

In [9]:
models[0]().fit(x_train, y_train , epochs =30 , batch_size=10 , verbose =2)

Epoch 1/30
12/12 - 0s - loss: 1.0014 - accuracy: 0.3750 - precision_4: 0.6364 - recall_4: 0.1167
Epoch 2/30
12/12 - 0s - loss: 0.8027 - accuracy: 0.6500 - precision_4: 0.7662 - recall_4: 0.4917
Epoch 3/30
12/12 - 0s - loss: 0.7096 - accuracy: 0.6917 - precision_4: 0.6506 - recall_4: 0.4500
Epoch 4/30
12/12 - 0s - loss: 0.6534 - accuracy: 0.6667 - precision_4: 0.6991 - recall_4: 0.6583
Epoch 5/30
12/12 - 0s - loss: 0.5653 - accuracy: 0.8417 - precision_4: 0.9412 - recall_4: 0.6667
Epoch 6/30
12/12 - 0s - loss: 0.5176 - accuracy: 0.9417 - precision_4: 0.9877 - recall_4: 0.6667
Epoch 7/30
12/12 - 0s - loss: 0.4953 - accuracy: 0.7167 - precision_4: 0.8081 - recall_4: 0.6667
Epoch 8/30
12/12 - 0s - loss: 0.4663 - accuracy: 0.9000 - precision_4: 0.9417 - recall_4: 0.8083
Epoch 9/30
12/12 - 0s - loss: 0.4384 - accuracy: 0.8833 - precision_4: 0.9048 - recall_4: 0.7917
Epoch 10/30
12/12 - 0s - loss: 0.4153 - accuracy: 0.9667 - precision_4: 0.9748 - recall_4: 0.9667
Epoch 11/30
12/12 - 0s - loss

In [10]:
print("Evaluate on test data")
models[0]().evaluate(x_test,y_test)

Evaluate on test data
1/1 [==============================] - 0s 998us/step - loss: 1.0269 - accuracy: 0.4333 - precision_5: 0.0000e+00 - recall_5: 0.0000e+00


[1.0269447565078735, 0.4333333373069763, 0.0, 0.0]

### Experiment 2 :

In [11]:
models[1]().fit(x_train, y_train , epochs =30 , batch_size=10 , verbose =2)

Epoch 1/30
12/12 - 0s - loss: 1.0031 - accuracy: 0.4167 - precision_6: 0.5000 - recall_6: 0.0500
Epoch 2/30
12/12 - 0s - loss: 0.8253 - accuracy: 0.6667 - precision_6: 0.7910 - recall_6: 0.4417
Epoch 3/30
12/12 - 0s - loss: 0.6979 - accuracy: 0.7417 - precision_6: 0.9828 - recall_6: 0.4750
Epoch 4/30
12/12 - 0s - loss: 0.6144 - accuracy: 0.8917 - precision_6: 1.0000 - recall_6: 0.5167
Epoch 5/30
12/12 - 0s - loss: 0.5651 - accuracy: 0.7167 - precision_6: 0.8316 - recall_6: 0.6583
Epoch 6/30
12/12 - 0s - loss: 0.5171 - accuracy: 0.9167 - precision_6: 0.9750 - recall_6: 0.6500
Epoch 7/30
12/12 - 0s - loss: 0.4854 - accuracy: 0.8417 - precision_6: 0.8788 - recall_6: 0.7250
Epoch 8/30
12/12 - 0s - loss: 0.4552 - accuracy: 0.8667 - precision_6: 0.8969 - recall_6: 0.7250
Epoch 9/30
12/12 - 0s - loss: 0.4341 - accuracy: 0.9667 - precision_6: 0.9652 - recall_6: 0.9250
Epoch 10/30
12/12 - 0s - loss: 0.4082 - accuracy: 0.9417 - precision_6: 0.9474 - recall_6: 0.9000
Epoch 11/30
12/12 - 0s - loss

In [12]:
print("Evaluate on test data")
models[1]().evaluate(x_test,y_test)

Evaluate on test data
1/1 [==============================] - 0s 1ms/step - loss: 1.2922 - accuracy: 0.3333 - precision_7: 0.5000 - recall_7: 0.3333


[1.2921665906906128, 0.3333333432674408, 0.5, 0.3333333432674408]

### Experiment 3 :

In [13]:
models[2]().fit(x_train, y_train , epochs =30 , batch_size=10 , verbose =2)

Epoch 1/30
12/12 - 0s - loss: 0.9461 - accuracy: 0.5167 - precision_8: 0.8000 - recall_8: 0.1667
Epoch 2/30
12/12 - 0s - loss: 0.6883 - accuracy: 0.7833 - precision_8: 0.8293 - recall_8: 0.5667
Epoch 3/30
12/12 - 0s - loss: 0.5224 - accuracy: 0.8667 - precision_8: 0.9135 - recall_8: 0.7917
Epoch 4/30
12/12 - 0s - loss: 0.4277 - accuracy: 0.8167 - precision_8: 0.8725 - recall_8: 0.7417
Epoch 5/30
12/12 - 0s - loss: 0.3599 - accuracy: 0.9583 - precision_8: 0.9658 - recall_8: 0.9417
Epoch 6/30
12/12 - 0s - loss: 0.3124 - accuracy: 0.9500 - precision_8: 0.9655 - recall_8: 0.9333
Epoch 7/30
12/12 - 0s - loss: 0.2804 - accuracy: 0.9500 - precision_8: 0.9500 - recall_8: 0.9500
Epoch 8/30
12/12 - 0s - loss: 0.2526 - accuracy: 0.9500 - precision_8: 0.9500 - recall_8: 0.9500
Epoch 9/30
12/12 - 0s - loss: 0.2261 - accuracy: 0.9667 - precision_8: 0.9667 - recall_8: 0.9667
Epoch 10/30
12/12 - 0s - loss: 0.2178 - accuracy: 0.9417 - precision_8: 0.9417 - recall_8: 0.9417
Epoch 11/30
12/12 - 0s - loss

In [14]:
print("Evaluate on test data")
models[2]().evaluate(x_test,y_test)

Evaluate on test data
1/1 [==============================] - 0s 718us/step - loss: 1.2382 - accuracy: 0.3333 - precision_9: 0.0000e+00 - recall_9: 0.0000e+00


[1.2381744384765625, 0.3333333432674408, 0.0, 0.0]

### Experiment 4

In [15]:
models[3]().fit(x_train, y_train , epochs =30 , batch_size=10 , verbose =2)

Epoch 1/30
12/12 - 0s - loss: 0.8340 - accuracy: 0.6583 - precision_10: 0.9048 - recall_10: 0.3167
Epoch 2/30
12/12 - 0s - loss: 0.5915 - accuracy: 0.7000 - precision_10: 0.8144 - recall_10: 0.6583
Epoch 3/30
12/12 - 0s - loss: 0.4700 - accuracy: 0.9417 - precision_10: 0.9712 - recall_10: 0.8417
Epoch 4/30
12/12 - 0s - loss: 0.4000 - accuracy: 0.8500 - precision_10: 0.8952 - recall_10: 0.7833
Epoch 5/30
12/12 - 0s - loss: 0.3461 - accuracy: 0.9500 - precision_10: 0.9500 - recall_10: 0.9500
Epoch 6/30
12/12 - 0s - loss: 0.3113 - accuracy: 0.9000 - precision_10: 0.8992 - recall_10: 0.8917
Epoch 7/30
12/12 - 0s - loss: 0.2808 - accuracy: 0.9167 - precision_10: 0.9160 - recall_10: 0.9083
Epoch 8/30
12/12 - 0s - loss: 0.2796 - accuracy: 0.9000 - precision_10: 0.9000 - recall_10: 0.9000
Epoch 9/30
12/12 - 0s - loss: 0.2579 - accuracy: 0.9000 - precision_10: 0.9000 - recall_10: 0.9000
Epoch 10/30
12/12 - 0s - loss: 0.2199 - accuracy: 0.9167 - precision_10: 0.9167 - recall_10: 0.9167
Epoch 11/

In [16]:
print("Evaluate on test data")
models[3]().evaluate(x_test,y_test)

Evaluate on test data
1/1 [==============================] - 0s 998us/step - loss: 1.2921 - accuracy: 0.3333 - precision_11: 0.4167 - recall_11: 0.3333


[1.2920950651168823,
 0.3333333432674408,
 0.4166666567325592,
 0.3333333432674408]

### Comparaison :

### Commentaire :
Pour les modéle proposés , la meilleure architecture est celle du 2éme modéle (au moment d'exécution) :<br>
train: accuracy: 0.9667 - precision: 0.9667 - recall: 0.9667<br>
test : accuracy: 0.6333 - precision: 0.0000 - recall: 0.0000<br>
2 couches :<br>
couche 1 : 256 neurones avec fonction d'activation relu <br>
couche 2 : 3 neurones avec fonction d'activation softmax <br>
fonction d'erreur : kl_divergence <br>
optimisation : Adam <br>
metriques : accuracy: 0.9867 - precision: 0.9867 - recall: 0.9867 

In [18]:
models[1]().summary()

Model: "model_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_33 (Dense)             (None, 256)               1280      
_________________________________________________________________
dense_34 (Dense)             (None, 3)                 771       
Total params: 2,051
Trainable params: 2,051
Non-trainable params: 0
_________________________________________________________________


## 3 Working towards the CNN
### Digit dataset

In [20]:
from sklearn.datasets import load_digits
(X_digits , y_digits) = load_digits(return_X_y=True , as_frame=True)

In [21]:
model = Sequential()
model.add(Dense(128 , activation='relu' , input_shape=(64,)))
# network.add(Dense(10, activation='relu', input_shape=(4,)))
model.add(Dense(128 , activation='relu'))
model.add(Dropout(0.2))
model.add(Dense(10 , activation='softmax'))
model.compile(optimizer='adam' , loss = 'categorical_crossentropy' , metrics=['accuracy',Precision(),Recall()])
model.fit(X_digits , to_categorical(y_digits) , epochs = 10  ,verbose = 2 )

Epoch 1/10
57/57 - 0s - loss: 1.4107 - accuracy: 0.6422 - precision_14: 0.7203 - recall_14: 0.5760
Epoch 2/10
57/57 - 0s - loss: 0.3074 - accuracy: 0.8993 - precision_14: 0.9220 - recall_14: 0.8753
Epoch 3/10
57/57 - 0s - loss: 0.1654 - accuracy: 0.9466 - precision_14: 0.9605 - recall_14: 0.9343
Epoch 4/10
57/57 - 0s - loss: 0.1229 - accuracy: 0.9605 - precision_14: 0.9673 - recall_14: 0.9538
Epoch 5/10
57/57 - 0s - loss: 0.0868 - accuracy: 0.9711 - precision_14: 0.9747 - recall_14: 0.9655
Epoch 6/10
57/57 - 0s - loss: 0.0624 - accuracy: 0.9789 - precision_14: 0.9832 - recall_14: 0.9766
Epoch 7/10
57/57 - 0s - loss: 0.0555 - accuracy: 0.9839 - precision_14: 0.9855 - recall_14: 0.9805
Epoch 8/10
57/57 - 0s - loss: 0.0482 - accuracy: 0.9866 - precision_14: 0.9905 - recall_14: 0.9844
Epoch 9/10
57/57 - 0s - loss: 0.0413 - accuracy: 0.9883 - precision_14: 0.9899 - recall_14: 0.9866
Epoch 10/10
57/57 - 0s - loss: 0.0276 - accuracy: 0.9928 - precision_14: 0.9939 - recall_14: 0.9911


### Faces Dataset

In [22]:
from sklearn.datasets import fetch_olivetti_faces
dataset = fetch_olivetti_faces(shuffle=True)

In [23]:
x = dataset.data
y = dataset.target

In [29]:
model = Sequential()
model.add(Dense(512 , activation='relu' , input_shape=(4096,)))
model.add(Dense(512 , activation='relu'))
model.add(Dense(40 , activation='softmax'))
model.compile(optimizer='adam' , loss = 'categorical_crossentropy' , metrics=['accuracy'])
model.fit(x , to_categorical(y) , epochs = 10  ,verbose = 2 )

Epoch 1/10
13/13 - 0s - loss: 4.4733 - accuracy: 0.0225
Epoch 2/10
13/13 - 0s - loss: 3.8322 - accuracy: 0.0275
Epoch 3/10
13/13 - 0s - loss: 3.6157 - accuracy: 0.0525
Epoch 4/10
13/13 - 0s - loss: 3.4822 - accuracy: 0.0850
Epoch 5/10
13/13 - 0s - loss: 3.2459 - accuracy: 0.1800
Epoch 6/10
13/13 - 0s - loss: 3.0323 - accuracy: 0.2100
Epoch 7/10
13/13 - 0s - loss: 2.8081 - accuracy: 0.2725
Epoch 8/10
13/13 - 0s - loss: 2.4662 - accuracy: 0.3775
Epoch 9/10
13/13 - 0s - loss: 2.1335 - accuracy: 0.4625
Epoch 10/10
13/13 - 0s - loss: 1.8170 - accuracy: 0.5775


In [25]:
model.summary()

Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_41 (Dense)             (None, 512)               2097664   
_________________________________________________________________
dense_42 (Dense)             (None, 512)               262656    
_________________________________________________________________
dense_43 (Dense)             (None, 512)               262656    
_________________________________________________________________
dense_44 (Dense)             (None, 40)                20520     
Total params: 2,643,496
Trainable params: 2,643,496
Non-trainable params: 0
_________________________________________________________________
